### PR TITLE
[SPARK-28939][SQL][2.4] Propagate SQLConf for plans executed by toRdd

### PR DESCRIPTION
**Pull request by mgaido91**
### What changes were proposed in this pull request?
    The PR proposes to create a custom `RDD` which enables to propagate `SQLConf` also in cases not tracked by SQL execution, as it happens when a `Dataset` is converted to and RDD either using `.rdd` or `.queryExecution.toRdd` and then the returned RDD is used to invoke actions on it.
    
In this way, SQL configs are effective also in these cases, while earlier they were ignored.
    
### Why are the changes needed?
    
Without this patch, all the times `.rdd` or `.queryExecution.toRdd` are used, all the SQL configs set are ignored. An example of a reproducer can be:
```
      withSQLConf(SQLConf.SUBEXPRESSION_ELIMINATION_ENABLED.key, "false") {
        val df = spark.range(2).selectExpr((0 to 5000).map(i => s"id as field_$i"): _*)
        df.createOrReplaceTempView("spark64kb")
        val data = spark.sql("select * from spark64kb limit 10")
        // Subexpression elimination is used here, despite it should have been disabled
        data.describe()
      }
```

### Why are the changes needed?
Without this patch, all the times `.rdd` or `.queryExecution.toRdd` are used, all the SQL configs set are ignored. An example of a reproducer can be:
```
      withSQLConf(SQLConf.SUBEXPRESSION_ELIMINATION_ENABLED.key, "false") {
        val df = spark.range(2).selectExpr((0 to 5000).map(i => s"id as field_$i"): _*)
        df.createOrReplaceTempView("spark64kb")
        val data = spark.sql("select * from spark64kb limit 10")
        // Subexpression elimination is used here, despite it should have been disabled
        data.describe()
      }
```

### Does this PR introduce any user-facing change?
When a user calls `.queryExecution.toRdd`, a `SQLExecutionRDD` is returned wrapping the `RDD` of the execute. When `.rdd` is used, an additional `SQLExecutionRDD` is present in the hierarchy.

### How was this patch tested?
added UT

### CODE REVIEW

**Comment by mgaido91** [comment]
cc @cloud-fan @hvanhovell @maropu @viirya 

I haven't yet added the flag requested by @hvanhovell as I have some doubts about it as I expressed in the other PR. I'll add if you think it is needed.

**Comment by dongjoon-hyun** [comment]
Hi, All.
Shall we hold on this? The original PR seems to break JDK11 build.
- https://github.com/apache/spark/actions

**Comment by dongjoon-hyun** [comment]
https://github.com/apache/spark/pull/25738 is ready.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecutionRDD.scala [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+ *
+ * @param sqlRDD the `RDD` generated by the SQL plan
+ * @param conf the `SQLConf` to apply to the execution of the SQL plan
+ */
+class SQLExecutionRDD(
+    var sqlRDD: RDD[InternalRow], conf: SQLConf) extends RDD[InternalRow](sqlRDD) {
+  private val sqlConfigs = conf.getAllConfs
+  private lazy val sqlConfExecutorSide = {
+    val props = new Properties()
+    props.putAll(sqlConfigs.asJava)
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecutionRDD.scala]: Hi, @mgaido91 .
Although Apache Spark `branch-2.4` doesn't support JDK11, some downstreams support JDK11 on `branch-2.4`.
Could you include #25738 to match `master` branch?

**Comment by mgaido91** [comment]
hi all, I am going to update this once the followups are done, meanwhile do you all agree adding the flag proposed by @hvanhovell ? If so, I'll add it when I update this PR. Thanks.

**Comment by dongjoon-hyun** [comment]
Yes. I'm +1 for @hvanhovell 's advice for the flag. Please add a document for that flag, too.
cc @gatorsmile

**Comment by maropu** [comment]
+1 for the flag to keep the legacy behaivour. Is that document @dongjoon-hyun suggested the migration guide? Yea, we need to update that, too.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -1298,6 +1298,14 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val USE_CONF_ON_RDD_OPERATION =
+    buildConf("spark.sql.legacy.rdd.applyConf")
+      .internal()
+      .doc("When false, SQL configurations are disregarded when operations on a RDD derived from" +
+        " a dataframe are executed. This is the (buggy) behavior up to 2.4.3.")
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: `up to 2.4.4` ?

File: docs/sql-migration-guide-upgrade.md [hunk 3]
```diff
@@ -7,6 +7,12 @@ displayTitle: Spark SQL Upgrading Guide
 * Table of contents
 {:toc}
 
+## Upgrading from Spark SQL 2.4 to 2.4.5
+
+ - Starting from 2.4.5, SQL configurations are effective also when a Dataset is converted to an RDD and its
+   plan is executed due to action on the derived RDD. The previous buggy behavior can be restored setting
+   `spark.sql.legacy.rdd.applyConf` to `false`.
```
Comment (dongjoon-hyun) [inline docs/sql-migration-guide-upgrade.md]: Hi, @hvanhovell , @cloud-fan and @gatorsmile . As @mgaido91 asked here, this PR will add this flag only at `branch-2.4`. In this case, is it okay this config will be **added and deprecated** at `2.4.5` and will be removed at `3.0.0`?

For me, we don't need to add this configuration to `master`. Did I understand correctly?

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to branch-2.4.
Sorry for being late, @mgaido91 .

cc @gatorsmile

**Pull request closed without merging**